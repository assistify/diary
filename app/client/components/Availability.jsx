import React from 'react';
import { PropTypes } from 'prop-types';
import Heading from 'react-bulma-components/lib/components/heading';
import Container from 'react-bulma-components/lib/components/container';
import Image from 'react-bulma-components/lib/components/image';
import Media from 'react-bulma-components/lib/components/media';
import Columns from 'react-bulma-components/lib/components/columns';
import User from './User';

export default function Availabilities(props) {
  const {
    members
  } = props;
  return (
    <Container className="c-availabilities">
      <Heading size={3}>Verf√ºgbarkeit</Heading>

      <Columns multiline centered>
        {
        members.map(member => (
          <Columns.Column size={3}>
            <Media>
              <Media.Item renderAs="figure" position="left">
                <Image renderAs="p" size={64} alt="64x64" src="http://bulma.io/images/placeholders/128x128.png" />
              </Media.Item>
              <Media.Item>
                <User username={member.username} />
                <p>{member.availability}</p>
              </Media.Item>
            </Media>
          </Columns.Column>
        ))
      }
      </Columns>
    </Container>
  );
}

Availabilities.propTypes = {
  members: PropTypes.arrayOf(PropTypes.shape({
    username: PropTypes.string.isRequired,
    availability: PropTypes.string.isRequired,
  })).isRequired
};
