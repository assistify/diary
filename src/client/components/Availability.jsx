import React from 'react';
import { PropTypes } from 'prop-types';
import Heading from 'react-bulma-components/lib/components/heading';
import Section from 'react-bulma-components/lib/components/section';
import Columns from 'react-bulma-components/lib/components/columns';
import Tag from 'react-bulma-components/lib/components/tag';
import UserFactsheet from './UserFactsheet';

import isBlocked from '../lib/isBlocked';
import BlockedIndication from './BlockedIndication';

export default function Availabilities(props) {
  const {
    members,
    contentEditable,
    updateValue
  } = props;

  return (
    <Section className="c-availabilities">
      <Heading size={3}>Verf√ºgbarkeit</Heading>

      <Columns multiline centered>
        {
          members
            .sort((a, b) => (a.statusKnown && !b.statusKnown ? -1
              : a.username > b.username))
            .map(member => (
              <Columns.Column
                key={member.username}
                size={4}
                className={member.statusKnown ? 'status-known' : 'status-unknown'}
              >
                <UserFactsheet
                  member={member}
                  contentEditable={contentEditable}
                  updateValue={updateValue}
                >
                  <Tag.Group>
                    {member.statusKnown && isBlocked(member.past.blockingItems)
                      && <BlockedIndication />
                    }
                    {member.statusKnown && member.future.availability
                      && (
                        <Tag
                          color="primary"
                          className="wrapped"
                          contentEditable={contentEditable}
                          suppressContentEditableWarning
                          onBlur={e => updateValue(member.username, 'future.availability', e.target.innerHTML)}
                        >
                          {member.future.availability}
                        </Tag>
                      )
                    }
                  </Tag.Group>
                </UserFactsheet>
              </Columns.Column>
            ))
        }
      </Columns>
    </Section>
  );
}

Availabilities.propTypes = {
  members: PropTypes.arrayOf(PropTypes.shape({
    username: PropTypes.string.isRequired,
    statusKnown: PropTypes.bool.isRequired,
    future: PropTypes.shape({
      availability: PropTypes.string,
      plannedItems: PropTypes.array
    }).isRequired,
    past: PropTypes.shape({
      completedItems: PropTypes.array,
      blockingItems: PropTypes.array
    }).isRequired
  })).isRequired,
  contentEditable: PropTypes.bool.isRequired,
  updateValue: PropTypes.func.isRequired
};
