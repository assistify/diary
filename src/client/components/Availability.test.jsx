import React from 'react';
import renderer from 'react-test-renderer';
import Availability from './Availability';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Availability
        members={[
          {
            username: 'user1',
            statusKnown: true,
            past: {
              completedItems: [],
              blockingItems: [{title: 'bla'}]
            },
            future: {
              availability: 'here',
              plannedItems: []
            }
          },
          {
            username: 'user2',
            statusKnown: false,
            past: {
              completedItems: []
            },
            future: {
              availability: '',
              plannedItems: []
            }
          },
        ]}
        contentEditable={false}
        updateValue={() => {}}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
