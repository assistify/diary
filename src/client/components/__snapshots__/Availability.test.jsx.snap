// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="section c-availabilities"
  style={Object {}}
>
  <h1
    className="title is-3"
  >
    Verf√ºgbarkeit
  </h1>
  <div
    className="columns is-multiline is-centered"
    style={Object {}}
  >
    <div
      className="status-known column is-4"
      style={Object {}}
    >
      <article
        className="media"
        style={Object {}}
      >
        <figure
          className="media-left"
          style={Object {}}
        >
          <figure
            className="image avatar is-128x128"
            style={Object {}}
          >
            <img
              alt="user1"
              onError={[Function]}
              src="/avatar/user1"
            />
          </figure>
        </figure>
        <div
          className="factSheetInfo"
        >
          <h1
            className="title is-5"
          >
            <a
              className="c-user"
              href="/direct/user1"
              rel="noopener noreferrer"
              target="_blank"
            >
              @user1
            </a>
          </h1>
          <span
            className="tags"
            style={Object {}}
          >
            <span
              className="tag blocked"
              style={Object {}}
            >
              Blockiert
            </span>
            <span
              className="tag wrapped is-primary"
              contentEditable={false}
              onBlur={[Function]}
              style={Object {}}
              suppressContentEditableWarning={true}
            >
              here
            </span>
          </span>
        </div>
      </article>
    </div>
    <div
      className="status-unknown column is-4"
      style={Object {}}
    >
      <article
        className="media"
        style={Object {}}
      >
        <figure
          className="media-left"
          style={Object {}}
        >
          <figure
            className="image avatar is-64x64"
            style={Object {}}
          >
            <img
              alt="user2"
              onError={[Function]}
              src="/avatar/user2"
            />
          </figure>
        </figure>
        <div
          className="factSheetInfo"
        >
          <h1
            className="title is-5"
          >
            <a
              className="c-user"
              href="/direct/user2"
              rel="noopener noreferrer"
              target="_blank"
            >
              @user2
            </a>
          </h1>
          <span
            className="tags"
            style={Object {}}
          />
        </div>
      </article>
    </div>
  </div>
</section>
`;
